rholang/src/test/scala/coop/rchain/rholang/ProcGen.scala

```diff
@@ -0,0 +1,247 @@
+      uri  <- uriGen
+    } yield new NameDeclUrn(name, uri)
+
+  lazy val nameDeclGen: Gen[NameDecl] = Gen.oneOf(nameDeclSimplGen, nameDeclUrnGen)
+
+  private def pnewGen(state: State): Gen[PNew] = {
+    val newState = state.decrementHeight
+
+    for {
+      seqNameDecl <- Gen.nonEmptyListOf(nameDeclSimplGen)
+      names       = extractNames(seqNameDecl)
+      proc        <- procGen(processContextProcs, newState.addNames(names))
+    } yield new PNew(seqToJavaCollection[ListNameDecl, NameDecl](seqNameDecl), proc)
+  }
+
+  lazy val bundleGen: Gen[Bundle] =
+    Gen.oneOf(new BundleWrite(), new BundleRead(), new BundleReadWrite(), new BundleEquiv())
+
+  private def procGen(procGens: Seq[State => Gen[Proc]], state: State): Gen[Proc] =
+    if (state.height > 0) {
+      oneOf(procGens.map(_.apply(state.decrementHeight)))
+    } else
+      pnilGen
+
+  private def allProcs: Seq[State => Gen[Proc]] =
+    Seq(pgroundGen, pparGen, psendGen, pnewGen, pvarGen)
+
+  private def processContextProcs: Seq[State => Gen[Proc]] =
+    Seq(pgroundGen, pparGen, psendGen, pnewGen)
+
+  def topLevelGen(height: Int): Gen[Proc] =
+    procGen(processContextProcs, State(height, Set.empty))
+
+  implicit def procShrinker: Shrink[Proc] = Shrink {
+    case p: PGround =>
+      (p.ground_ match {
+        case p: GroundInt =>
+          shrinkIntegral[Long]
+            .shrink(p.longliteral_.toLong)
+            .map(n => new GroundInt(n.toString))
+        case p: GroundBool => streamSingleton(p)
+        case p: GroundString =>
+          shrinkString
+            .shrink(withoutQuotes(p.stringliteral_))
+            .map(s => new GroundString(withQuotes('"')(s)))
+        case p: GroundUri =>
+          val components = withoutQuotes(p.uriliteral_).split(":")
+
+          for {
+            shrinkedComponentSeq <- shrinkContainer[Seq, String]
+                                     .shrink(components)
+                                     .takeWhile(_.nonEmpty)
+            shrinkedComponents <- shrinkedComponentSeq.map(c => shrinkString.shrink(c))
+          } yield new GroundUri(mkGroundUri(shrinkedComponents))
+      }).map(new PGround(_))
+
+    case p: PPar =>
+      for {
+        sp1 <- shrink(p.proc_1)
+        sp2 <- shrink(p.proc_2)
+      } yield new PPar(sp1, sp2)
+
+    case p: PSend =>
+      val initialProcs = javaCollectionToSeq[ListProc, Proc](p.listproc_)
+
+      shrinkContainer[Seq, Proc]
+        .shrink(initialProcs)
+        .takeWhile(_.nonEmpty)
+        .map(procs => new PSend(p.name_, p.send_, seqToJavaCollection[ListProc, Proc](procs)))
+
+    case p: PNew =>
+      val initialNames = javaCollectionToSeq[ListNameDecl, NameDecl](p.listnamedecl_)
+
+      for {
+        newNames     <- shrinkContainer[Seq, NameDecl].shrink(initialNames).takeWhile(_.nonEmpty)
+        newLocalVars = extractNames(newNames)
+        proc <- shrink(p.proc_)
+                 .filter(p => extractFreeVariables(p).forall(newLocalVars.toSet.contains(_)))
```

Comment:
move `toSet` out of the loop (to newLocalVars)

Reply:
`newLocalVars.conainsAll(extractFreeVariables(p))` ?

Reply:
actually extractNames returns a `Set` so the call to `toSet` is not needed

Reply:
there is no method `Set.containsAll`

Reply:
subsetOf